File: Programs/Total_Characters_in_String_After_Transformations_II.cpp
class Solution {
    const int N = 26;
    const int MOD = 1e9 + 7;

    using Matrix = vector<vector<int>>;
    Matrix createMatrix(const int& m, const int& n, const int val = 0) {
        return Matrix(m, vector<int>(n, val));
    }

    Matrix multiplyMatrix(const Matrix& A, const Matrix& B) {
        int m = A.size(), n = A[0].size(), p = B[0].size();
        Matrix C = createMatrix(m, p, 0);

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < p; ++j) {
                for (int k = 0; k < n; ++k) {
                    C[i][j] = (C[i][j] + 1LL * A[i][k] * B[k][j]) % MOD;
                }
            }
        }

        return C;
    }
    void powerMatrix(Matrix& A, int n) {
        const int m = A.size();
        Matrix res = createMatrix(m, m, 0);
        for (int i = 0; i < m; ++i) res[i][i] = 1;

        while (n > 0) {
            if (n & 1) res = multiplyMatrix(res, A);

            A = multiplyMatrix(A, A);
            n >>= 1;
        }

        A = res;
    }

public:
    int lengthAfterTransformations(string str, int t, vector<int>& nums) {
        Matrix freq = createMatrix(N, 1, 0);
        for (const char& ch : str) ++freq[ch - 'a'][0];

        Matrix T = createMatrix(N, N, 0);
        for (int c = 0; c < 26; ++c) {
            for (int i = 0; i < nums[c]; ++i) {
                int c_dash = (c + i + 1) % N;
                T[c_dash][c] = 1;
            }
        }

        powerMatrix(T, t);

        Matrix ans = multiplyMatrix(T, freq);

        int count = 0;
        for (int i = 0; i < 26; ++i) count = (count + ans[i][0]) % MOD;
        return count;
    }
};

File: Programs/Total_Characters_in_String_After_Transformations_II.py
# Without Numpy
class Solution:
    def __init__(self):
        self.N= 26
        self.MOD= int(1e9 + 7)

    def createMatrix(self, m, n, val= 0):
        return [[val for _ in range(n)] for _ in range(m)]
    
    def multiplyMatrix(self, A, B):
        m, n, p= len(A), len(A[0]), len(B[0])
        C= self.createMatrix(m, p, 0)

        for i in range(m):
            for j in range(p):
                for k in range(n):
                    C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % self.MOD
        
        return C
    def powerMatrix(self, A, t):
        m= len(A)
        res= self.createMatrix(m, m)
        for i in range(m):
            res[i][i]= 1
        
        while t > 0:
            if t & 1:
                res= self.multiplyMatrix(res, A)
            
            A= self.multiplyMatrix(A, A)
            t >>= 1
        
        return res

    def lengthAfterTransformations(self, s: str, t: int, nums: List[int]) -> int:
        freq= self.createMatrix(self.N, 1, 0)
        for ch in s:
            freq[ord(ch) - ord('a')][0] += 1
        
        T= self.createMatrix(self.N, self.N, 0)
        for c in range(self.N):
            for i in range(nums[c]):
                c_dash= (c + i + 1) % self.N
                T[c_dash][c]= 1
        
        T= self.powerMatrix(T, t)
        freq= self.multiplyMatrix(T, freq)

        count= 0
        for i in range(self.N):
            count= (count + freq[i][0]) % self.MOD
        return count
    
# With Numpy
import numpy as np

class Solution:
    def __init__(self):
        self.N= 26
        self.MOD= int(1e9 + 7)

    def powerMatrix(self, A, power):
        m= len(A)
        res= np.identity(m, dtype= np.int64)
        
        while power > 0:
            if power & 1:
                res= np.dot(res, A) % self.MOD
            
            A= np.dot(A, A) % self.MOD
            power >>= 1
        
        return res

    def lengthAfterTransformations(self, s: str, t: int, nums: List[int]) -> int:
        freq= np.zeros((self.N, 1), dtype= np.int64)
        for ch in s:
            freq[ord(ch) - ord('a')][0] += 1
        
        T= np.zeros((self.N, self.N), dtype= np.int64)
        for c in range(self.N):
            for i in range(nums[c]):
                c_dash= (c + i + 1) % self.N
                T[c_dash][c]= 1
        
        T= self.powerMatrix(T, t)
        freq= np.dot(T, freq) % self.MOD

        return int(np.sum(freq) % self.MOD)